zebra-network/src/address_book.rs

```diff
@@ -5,54 +5,73 @@ use std::{
     collections::{BTreeSet, HashMap},
     iter::Extend,
     net::SocketAddr,
+    sync::Arc,
 };
 
 use chrono::{DateTime, Utc};
 use tracing::Span;
 
-use crate::{
-    constants,
-    types::{MetaAddr, PeerServices},
-};
+use crate::{constants, types::MetaAddr, PeerConnectionState};
 
 /// A database of peers, their advertised services, and information on when they
 /// were last seen.
 #[derive(Debug)]
 pub struct AddressBook {
-    by_addr: HashMap<SocketAddr, (DateTime<Utc>, PeerServices)>,
-    by_time: BTreeSet<MetaAddr>,
+    /// Maps each peer address to its matching `MetaAddr`
+    by_addr: HashMap<SocketAddr, Arc<MetaAddr>>,
+
+    /// Stores peer addresses, in state then reconnection order.
+    ///
+    /// This order should not be used for reconnection attempts: use
+    /// [`reconnection_peers`] instead.
+    by_state: BTreeSet<Arc<MetaAddr>>,
+
+    /// The span for operations on this address book.
```

Comment:
Open a follow-up ticket to set up a persistent peer address book cache using RocksDB?